File: testdata/precision-suite-cpp/tp_pathtraversal.c
// Path traversal: a user-controlled path read from the environment is passed
// straight to fopen(3) with no realpath canonicalization or base-prefix check,
// so `../../etc/passwd` escapes the intended directory. This is CWE-22. A correct
// scanner fires TAINT-004.
#include <stdio.h>
#include <stdlib.h>

FILE *open_report(void) {
    char *path = getenv("REPORT_PATH");
    FILE *f = fopen(path, "r"); // nox-expect: TAINT-004
    return f;
}

File: testdata/precision-suite-rust/tp_pathtraversal.rs
// Path traversal: a user-controlled path is passed straight to std::fs::read
// with no canonicalization or file_name() component-stripping, so `../../etc/
// passwd` escapes the intended directory. This is CWE-22. A correct scanner
// fires TAINT-004.
use std::env;
use std::fs;

fn serve_file() -> Vec<u8> {
    let user_path = env::var("FILE").unwrap_or_default();
    let data = fs::read(user_path).unwrap_or_default(); // nox-expect: TAINT-004
    data
}
